File: python/quick_sort.py
import numpy as np
import random
from pandas import pivot
import pytest


def randomized_partition(nums, l, r):
    pivot = random.randint(l, r)
    nums[pivot], nums[r] = nums[r], nums[pivot]
    i = l - 1
    for j in range(l, r):
        if nums[j] < nums[r]:
            i += 1
            nums[i], nums[j] = nums[j], nums[i]
    nums[i + 1], nums[r] = nums[r], nums[i + 1]
    return i + 1

def quick_sort(nums, l, r):  # O(nlog(n))/O(n)
    if l < r:
        mid = randomized_partition(nums, l, r)
        quick_sort(nums, l, mid - 1)
        quick_sort(nums, mid + 1, r)

    return nums


def test_sort():
    for i in range(1000):
        nums = list(np.random.randint(0, 1000, size=50))
        assert sorted(nums) == quick_sort(nums, 0, len(nums) - 1)
    return


if __name__ == '__main__':
    random.seed(2023)
    np.random.seed(2023)
    nums = list(np.random.randint(0, 20, size=10))
    print(nums)
    print(quick_sort(nums, 0, len(nums) - 1))

File: cpp/quick_sort.cpp
#include<iostream>
#include<vector>
#include<stdlib.h>
#include <time.h>

using namespace std;

int randomized_partition(vector<int>& nums, int l, int r)
{
    int pivot = l + rand() % (r - l + 1);
    int i = l - 1;
    for(int j = l; j < r; ++j){
        if(nums[j] < nums[r]){
            i++;
            swap(nums[i], nums[j]);
        }
    }
    swap(nums[i + 1], nums[r]);
    return i + 1;
}

vector<int> quick_sort(vector<int> &nums, int l, int r)
{
    if(l < r){
        int mid = randomized_partition(nums, l, r);
        quick_sort(nums, l, mid - 1);
        quick_sort(nums, mid + 1, r);
    }
    return nums;
}

int main()
{
    srand((unsigned)time(0));
    vector<int> nums;
    for (int i = 0; i < 10; ++i)
    {
        nums.push_back(rand() % 20);
    }

    for (int i = 0; i < nums.size(); ++i) 
    {
        cout << nums.at(i) << ' ';
    }

    cout << endl;

    nums = quick_sort(nums, 0, nums.size() - 1);
    
    for (int i = 0; i < nums.size(); ++i) 
    {
        cout << nums.at(i) << ' ';
    }
    
    return 0;
}